### PR TITLE
feat(#540): Support for registry auth credentials from env var DOCKER_AUTH_CONFIG

### PR DESCRIPTION
**Pull request by vova-lantsov-dev**

### CODE REVIEW

**Comment by vova-lantsov-dev** [comment]
@HofmeisterAn ubuntu pipeline got stuck?

File: src/Testcontainers/Builders/EnvironmentVariableBase64Provider.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    private readonly ILogger logger;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EnvironmentVariableBase64Provider"/> class.
+    /// </summary>
+    /// <param name="logger">The logger.</param>
+    [PublicAPI]
+    public EnvironmentVariableBase64Provider(ILogger logger)
+    {
+      var environmentVariableValue = Environment.GetEnvironmentVariable("DOCKER_AUTH_CONFIG");
```
Comment (HofmeisterAn) [inline src/Testcontainers/Builders/EnvironmentVariableBase64Provider.cs]: Can we move this to `CustomConfiguration`? Something like `public JsonElement GetDockerAuthConfig()`? `EnvironmentConfiguration` and `PropertiesFileConfiguration` will be singeltons to access the configuration in the future.

File: src/Testcontainers/Builders/DockerRegistryAuthenticationProvider.cs [hunk 2]
```diff
@@ -84,7 +84,11 @@ private IDockerRegistryAuthenticationConfiguration GetUncachedAuthConfig(string
           {
             authConfig = new IDockerRegistryAuthenticationProvider[]
               {
-                new CredsHelperProvider(dockerConfigDocument, this.logger), new CredsStoreProvider(dockerConfigDocument, this.logger), new Base64Provider(dockerConfigDocument, this.logger),
+                // environment variable provider goes first as it should overwrite the Docker configuration file when set
+                new EnvironmentVariableBase64Provider(this.logger),
+                new CredsHelperProvider(dockerConfigDocument, this.logger),
+                new CredsStoreProvider(dockerConfigDocument, this.logger),
+                new Base64Provider(dockerConfigDocument, this.logger),
```
Comment (HofmeisterAn) [inline src/Testcontainers/Builders/DockerRegistryAuthenticationProvider.cs:97]: If you extend `CustomConfiguration` as suggested above, you can pass the `GetDockerAuthConfig()` result to `Base64Provider`. I think you don't need `EnvironmentVariableBase64Provider` at all.

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
Thanks for your pull request. I have two minor suggestions. I really like your `EnvironmentVariableScope` approach, nice idea.

> @HofmeisterAn ubuntu pipeline got stuck?

Yep, since one of the last pull request the pipeline dose not run reliable anymore 😕. I am on it.

**Comment by vova-lantsov-dev** [comment]
@HofmeisterAn I don't completely understand the purpose of these classes (`CustomConfiguration`, `EnvironmentConfiguration` and `PropertiesFileConfiguration`) in case of `DockerRegistryAuthenticationProvider` logic, but I made some changes, please review them and let me know if I misunderstood anything

**Review by HofmeisterAn — COMMENTED** [review]
> @HofmeisterAn I don't completely understand the purpose of these classes (`CustomConfiguration`, `EnvironmentConfiguration` and `PropertiesFileConfiguration`) in case of `DockerRegistryAuthenticationProvider` logic, but I made some changes, please review them and let me know if I misunderstood anything

I was thinking about something like this: [vova-lantsov-dev-feature/env-var-base64-auth](https://github.com/testcontainers/testcontainers-dotnet/compare/vova-lantsov-dev-feature/env-var-base64-auth?expand=1). I have not tested it yet, but what do you think?

- [x] Add Docker auth config test to `CustomConfigurationTest`

**Review by HofmeisterAn — COMMENTED** [review]
If you have some time to test [vova-lantsov-dev-feature/env-var-base64-auth](https://github.com/testcontainers/testcontainers-dotnet/tree/vova-lantsov-dev-feature/env-var-base64-auth), I can merge your and my changes and close this PR.

**Comment by vova-lantsov-dev** [comment]
I reviewed and applied your suggestions, but with small change:

image

`GetDefaultDockerAuthConfig()` now returns nullable value because we want to check for default docker config even if `dockerConfigFile` doesn't exist (`else if` block)

**Comment by vova-lantsov-dev** [comment]
I can test this on my local machine and CI server if you pack it into the NuGet

**Review by HofmeisterAn — APPROVED** [review]

**Review by HofmeisterAn — COMMENTED** [review]
Published in `2.2.0-beta.2835065501` (ced98a9db6c70c188b113938decc25d12d03e6fc), closes #540.

**Comment by vova-lantsov-dev** [comment]
@HofmeisterAn I confirm that it works locally with `.docker` directory and on CI/CD server with `DOCKER_AUTH_CONFIG` env variable. I think we can merge this to develop and pack into release tag

**Comment by vova-lantsov-dev** [comment]
By the way, it looks like I tested not the same code as this PR produces. That's why we have conflict

**Comment by vova-lantsov-dev** [comment]
Oh, I see, you made some further changes. Do you want me to repeat them in my code, or you will just close the PR?

**Comment by HofmeisterAn** [comment]
> Do you want me to repeat them in my code, or you will just close the PR?

No, not necessary. I merged and pushed all changes to develop already. You tested the right version. Thanks again.

**Comment by vova-lantsov-dev** [comment]
Hi @HofmeisterAn 
Please let me know when you plan to publish the release version containing this feature? It's quite strange to use the beta version of package in production environment :)

**Comment by HofmeisterAn** [comment]
No ETA yet. I'd like to add the custom HTTP headers to the next version first.

**Pull request closed without merging**